### ISSUE TITLE
[REQUEST] Bindset configuration

### ISSUE THREAD

**Issue by hissssst** [mawww/kakoune#4836]
### Feature

By the word `bindset` I mean a set of default bindings.

For example, in `.kakrc` we could have
```kakoune
default-bindset pre-v2022.10.31
```
And this configuration would have `,` as a user command mode and `<space>` as a command to leave one cursor instead of multiple.

And for `.kakrc` with
```kakoune
default-bindset v2022.10.31
```
We would have modern defaults.

### Usecase

With `default-bindset` it is possible to provide breaking changes to default bindings without actually breaking anyone's workflow.

This approach with specifying the version of default values is widely adopted in different configuration systems. For example, NixOS's `stateVersion` behaves this way and prevents breaking changes when updating configuration

**Comment by arrufat** [comment]
As someone who ported the golf challenges in https://github.com/mawww/golf/pull/37, it does not seem obvious to me how you would update all the kak scripts shipped with kakoune automatically when switching between binders (unless you kept a copy of all the kak scripts)

**Comment by hissssst** [comment]
The reason here is that you, as a user of kakoune, won't need to update anything. That's the main point of this feature request

**Comment by arrufat** [comment]
Yes, I got it, I was thinking about the implementation details of this feature request so that it's transparent to the user. I have the feeling that it complicates the editor a bit, which goes against its main design goal. Just my two cents.

**Comment by hissssst** [comment]
> @arrufat  complicates the editor a bit

How?

**Comment by arrufat** [comment]
> How?

By having to maintain several sets of key bindings and their corresponding kak scripts. It's going to be a burden for developers.

I guess you're suggesting this feature so that breaking changes do not break users workflows.
But it also lowers the barrier to introduce breaking changes, since there's a "solution" now.

This means that all scripts should have versions for each corresponding bindset. Some changes can be automated by search and replace, but some others may not (like the fact that p/P now select the pasted text). So in order to contribute/modify/improve a kak script, you'd have to do it as many times as bindsets exist, and make sure they all work properly.

**Comment by hissssst** [comment]
> maintain several sets of key bindings

That's not how it should work. I'd expect the developers to maintain only latest set of keybindings, and have functions which can restore previous behaviour of bindings.

> But it also lowers the barrier to introduce breaking changes, since there's a "solution" now.

That's not a correct point. First of all, there  is already no barrier for introducing breaking changes as we can already see. Second, breaking changes occur, and this happens in lifecycle of every project. And this is the project of a single person, thus decision about breaking changes is not based on statistics or voting.

> This means that all scripts should have versions for each corresponding bindset

I don't think that's how it should be implemented.

**Comment by arrufat** [comment]
Maybe by adding a version of the bindset at the beginning of each script it could be feasible. 

However, it seems like asking the developers to maintain a fork for you. And also to never be able to get rid of bad or outdated choices they might have made in the past.

**Comment by hissssst** [comment]
> However, it seems like asking the developers to maintain a fork for you. And also to never be able to get rid of bad or outdated choices they might have made in the past.

You're talking about your own idea of implementing this request, and this critics apply only to your idea.

Let me explain what kind of implementation I have in mind:

This command
```
default-bindset pre-v2022.10.31
```

to be translated to something like
```
default-bindset v2022.10.31

map global normal , <space>
map global normal <space> ,

map global normal X :_old_X_behavior<ret>
```
Where these `map`'s affect all other mappings.

> Edit: I don't agree with the fact that there's no barrier

And I don't agree that adding backward compatibility option is lowering this barrier

**Comment by arrufat** [comment]
Yes, I got that, my question is how would you handle all the scripts that are in in the `rc` directory and rely on those mappings?

**Comment by hissssst** [comment]
> how would you handle all the scripts that are in the rc directory

Execute them in latest bindset. As far as I know, they're executed before user's `.kakrc`, right?

**Comment by arrufat** [comment]
Then you should make sure to also load all plug-ins before you change the mappings in your kakrc

**Comment by Screwtapello** [comment]
Commands run with `execute-keys` are not affected by user mappings, so when the "default" mappings are updated, the rc scripts would be updated to match and then everything would work.

One problem with this proposal is that there is no sequence of current normal-mode commands that exactly replicates the previous behaviour of X. That was largely the point: the previous behaviour of X was stateful, and it was impossible to inspect or update that state other than by using X to see what would happen. The new behaviours are not stateful, which makes them more useful for scripting and more predictable to learn, but also makes it impossible to replicate the statefulness of the old behaviour.

**Comment by arrufat** [comment]
@Screwtapello I was about to say that about exec. How about source, though? It does not have a -with-maps switch, does this mean it always ignores mappings?

**Comment by hissssst** [comment]
> @Screwtapello Commands run with execute-keys are not affected by user mappings, so when the "default" mappings are updated, the rc scripts would be updated to match and then everything would work.

That's why I am asking for a feature request. If this was not the problem, I could've just created repository with compatibility commands.

> @Screwtapello One problem with this proposal is that there is no sequence of current normal-mode commands that exactly replicates the previous behaviour of X. That was largely the point: the previous behaviour of X was stateful, and it was impossible to inspect or update that state other than by using X to see what would happen. The new behaviours are not stateful, which makes them more useful for scripting and more predictable to learn, but also makes it impossible to replicate the statefulness of the old behaviour.

Again, that's why I am asking for feature request. I know about this problem, and I am already using not completely correct bindmap for old X behavior.
